omegaconf/_utils.py

```diff
@@ -237,6 +237,23 @@ def get_attr_data(obj: Any, allow_objects: Optional[bool] = None) -> Dict[str, A
         except ValidationError as ex:
             format_and_raise(node=None, key=name, value=value, cause=ex, msg=str(ex))
         d[name]._set_parent(None)
+    if is_dict_subclass(obj_type) and not isinstance(obj, type):
+        key_type, element_type = get_dict_key_value_types(obj_type)
+        for name, value in obj.items():
+            is_optional, type_ = _resolve_optional(element_type)
+            type_ = _resolve_forward(type_, obj.__module__)
+            try:
+                d[name] = _maybe_wrap(
+                    ref_type=type_,
+                    is_optional=is_optional,
+                    key=name,
+                    value=value,
+                    parent=dummy_parent,
+                )
+            except ValidationError as ex:
+                format_and_raise(
+                    node=None, key=name, value=value, cause=ex, msg=str(ex)
+                )
```

Comment:
Looks identical to get_dataclass_data, can you refactor to use a common function for this block?

Reply:
1. What is the purpose of `not isinstance(obj, type)` in the if?
2. It feels like DictConf shoudl correctly extract the fields from it in the primary loop inside _maybe_wrap(), did you look to see why it's not happening? (I didn't run debug this so I may be wrong).

Reply:
> Looks identical to get_dataclass_data, can you refactor to use a common function for this block?

Ok, will do.

Reply:
>     1. What is the purpose of `not isinstance(obj, type)` in the if?

This way, we only get class instances, not classes.

For example `isinstance(MyDataclass(), type)` is false but `isinstance(MyDataclass, type)` is true.

Calling `instance.items()` makes sense, but calling `MyDataclass.items()` does not work.

Reply:
>     2. It feels like DictConf shoudl correctly extract the fields from it in the primary loop inside _maybe_wrap(), did you look to see why it's not happening? (I didn't run debug this so I may be wrong).

It's because `obj` does not get passed to `_maybe_wrap`; only the fields/values are getting passed to `_maybe_wrap`, one at a time.

Reply:
> Looks identical to get_dataclass_data, can you refactor to use a common function for this block?

Refactored in https://github.com/omry/omegaconf/pull/653/commits/d30c7d70e92df32ee0b3e61327f38e7d136f7f9c.